src/wp-includes/functions.php

```diff
@@ -8458,3 +8458,87 @@ function wp_recursive_ksort( &$array ) {
 	}
 	ksort( $array );
 }
+
+/**
+ * Attempt to detect a VirtualBox environment.
+ *
+ * This attempts all known methods of detecting VirtualBox.
+ *
+ * @since 6.1.0
+ *
+ * @global WP_Filesystem_Base $wp_filesystem WordPress filesystem subclass.
+ *
+ * @return bool Whether or not VirtualBox was detected.
+ */
+function is_virtualbox() {
+	global $wp_filesystem;
+	static $is_virtualbox;
+
+	if ( null !== $is_virtualbox ) {
+		return $is_virtualbox;
+	}
+
+	/**
+	 * Filters whether the current environment uses VirtualBox.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @param bool Whether the current environment uses VirtualBox. Default false.
+	 */
+	if ( apply_filters( 'is_virtualbox', false ) ) {
+		$is_virtualbox = true;
+		return $is_virtualbox;
+	}
+
+	// Detection via Composer.
+	if ( function_exists( 'getenv' ) && 'virtualbox' === getenv( 'COMPOSER_RUNTIME_ENV' ) ) {
+		$is_virtualbox = true;
+		return $is_virtualbox;
+	}
+
+	$virtualbox_unames = array( 'vvv' );
+
+	// Detection via `php_uname()`.
+	if ( function_exists( 'php_uname' ) && in_array( php_uname( 'n' ), $virtualbox_unames, true ) ) {
+		$is_virtualbox = true;
+		return $is_virtualbox;
+	}
+
+	/*
+	 * Vagrant can use alternative providers.
+	 * This may not be reliable without some additional checks.
+	 */
+	$virtualbox_usernames = array( 'vagrant' );
+
+	// Detection via user name with POSIX.
+	if ( function_exists( 'posix_getpwuid' ) && function_exists( 'posix_geteuid' ) ) {
+		$user = posix_getpwuid( posix_geteuid() );
+		if ( $user && in_array( $user['name'], $virtualbox_usernames, true ) ) {
+			$is_virtualbox = true;
+			return $is_virtualbox;
+		}
+	}
+
+	// Initialize the filesystem if not set.
+	if ( ! $wp_filesystem ) {
+		require_once ABSPATH . '/wp-admin/includes/file.php';
+		WP_Filesystem();
+	}
+
+	// Detection via file owner.
+	if ( in_array( $wp_filesystem->owner( __FILE__ ), $virtualbox_usernames, true ) ) {
+		$is_virtualbox = true;
+		return $is_virtualbox;
+	}
+
+	// Detection via file group.
+	if ( in_array( $wp_filesystem->group( __FILE__ ), $virtualbox_usernames, true ) ) {
+		$is_virtualbox = true;
+		return $is_virtualbox;
+	}
+
+	// Give up.
+	$is_virtualbox = false;
```

Comment:
```php
// Give up.
```

This is basically why I am concerned about this function and have been since it was first mentioned in Slack and I explained I wasn't keen on the idea.

Reply:
This function exists due to the constant/environment variable approach being blocked and a programmatic way of detecting VirtualBox being requested.

Without introducing a new library to Core, I believe that the detection methods in this function are basically the best available to us.

In terms of detecting VirtualBox to avoid this VirtualBox-specific bug, I still feel that the current most reliable way is a constant/environment variable.

In the event that the VirtualBox bug gets fixed, the constant/environment variable simply becomes useless, but removing references to it in Core at that stage won't break BC. For extenders, they could continue to use/require the constant/environment variable if it's necessary for them for some other reason.

So far, all other attempts to overcome this bug have either failed to work, or to get approval, and I'm currently out of ideas.

I can only ask that feedback might come with suggested paths forward so that we have a clear direction for next steps.